### PR TITLE
Implement Username/Email Login Identifier

### PR DESCRIPTION
**Pull request by solverat**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Fixed tickets | #1290 

See #1290 for discussion.

### CODE REVIEW

File: docs/03_Development/12_Customers/04_Registration_Types.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+```yaml
+core_shop_customer:
+    login_identifier: 'username'
+```
+
+## Security
+
+### Form (Frontend)
+CoreShop comes with a preinstalled constraints which will tell your customer, if an email address or username - depending on your settings - is valid or not.
```
Comment (jdreesen) [inline docs/03_Development/12_Customers/04_Registration_Types.md]:
```suggestion
CoreShop comes with a preinstalled constraint which will tell your customer, if an email address or username - depending on your settings - is valid or not.
```

File: docs/03_Development/12_Customers/04_Registration_Types.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+### Backend / API
+Plus, if you're going to update a customer by API or Backend, coreshop also checks if you're customer entity has unique data.
```
Comment (jdreesen) [inline docs/03_Development/12_Customers/04_Registration_Types.md]:
```suggestion
Plus, if you're going to update a customer by API or Backend, coreshop also checks if your customer entity has unique data.
```

File: docs/03_Development/12_Customers/04_Registration_Types.md [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+> Note: Both checks only applies to non-guests entities!
```
Comment (jdreesen) [inline docs/03_Development/12_Customers/04_Registration_Types.md]:
```suggestion
> Note: Both checks only apply to non-guest entities!
```

File: src/CoreShop/Bundle/ResourceBundle/Validator/Constraints/UniqueEntityValidator.php [hunk 4]
```diff
@@ -15,13 +15,58 @@
...
+     * @var ContainerInterface
+     */
+    protected $container;
+
+    /**
+     * @param ContainerInterface $container
+     */
+    public function __construct(ContainerInterface $container)
+    {
+        $this->expressionLanguage = new ExpressionLanguage();
```
Comment (dpfaffenbauer) [inline src/CoreShop/Bundle/ResourceBundle/Validator/Constraints/UniqueEntityValidator.php]: use CoreShop Expression Service

**Pull request merged**